src/sofa/pbrpc/web_service.cc

```diff
@@ -0,0 +1,539 @@
+        _servlet_map.swap(servlet_map);
+    }
+    std::pair<ServletMap::iterator, bool> ret = 
+        _servlet_map->insert(std::make_pair(real_path, servlet));
+    if (ret.second == true)
+    {
+#if defined( LOG )
+        LOG(INFO) << "RegisterServlet(): register webserver path {" 
+            << path << "} format to {" << real_path << "} success";
+#else
+        SLOG(INFO, "RegisterServlet(): register webserver path {%s} format to {%s} success", 
+             path.c_str(), real_path.c_str());
+#endif
+    }
+    else
+    {
+#if defined( LOG )
+        LOG(ERROR) << "RegisterServlet(): register webserver path {" 
+            << path << "} format to {" << real_path << "} alreay exist";
+#else
+        SLOG(ERROR, "RegisterServlet(): register webserver path {%s} format to {%s} already exist", 
+             path.c_str(), real_path.c_str());
+#endif
+    }
+    return ret.second;
+}
+
+bool WebService::UnregisterServlet(const std::string& path)
+{
+    std::string real_path = FormatPath(path);
+    ScopedLocker<FastLock> _(_servlet_map_lock);
+    if (!_servlet_map.unique())
+    {
+        ServletMapPtr servlet_map(new ServletMap(*_servlet_map));
+        _servlet_map.swap(servlet_map);
+    }
+    ServletMap::iterator it = _servlet_map->find(real_path);
+    if (it == _servlet_map->end())
+    {
+#if defined( LOG )
+        LOG(ERROR) << "UnregisterServlet(): unregister webserver path {" 
+            << path << "} format to {" << real_path << "} not exist";
+#else
+        SLOG(ERROR, "UnregisterServlet(): unregister webserver path {%s} format to {%s} not exist", 
+             path.c_str(), real_path.c_str());
+#endif
+        return false;
+    }
+    _servlet_map->erase(it);
+#if defined( LOG )
+    LOG(INFO) << "UnregisterServlet(): unregister webserver path {" 
+        << path << "} format to {" << real_path << "} success";
+#else
+    SLOG(INFO, "UnregisterServlet(): unregister webserver path {%s} format to {%s} success", 
+         path.c_str(), real_path.c_str());
+#endif
+    return true;
+}
+
+bool WebService::RoutePage(
+    const RpcRequestPtr& rpc_request,
+    const RpcServerStreamWPtr& server_stream)
+{
+    const HTTPRpcRequest* http_rpc_request = 
+        static_cast<const HTTPRpcRequest*>(rpc_request.get());
+
+    HTTPRequest request;
+    request.path = http_rpc_request->_path;
+    request.body = http_rpc_request->_req_body->ToString();
+    request.headers = http_rpc_request->_headers;
+    request.query_params = http_rpc_request->_query_params;
+
+    HTTPResponse response;
+    response.ip = HostOfRpcEndpoint(http_rpc_request->_local_endpoint);
+    response.port = PortOfRpcEndpoint(http_rpc_request->_local_endpoint);
+    response.host = GetHostName(response.ip);
+
+    bool ret = false;
+    const std::string& method = http_rpc_request->_method;
+    Servlet servlet = FindServlet(method);
+    if (servlet)
+    {
+        ret = servlet->Run(request, response);
```

Comment:
if ret = false, the server will response "method not found" to client, is it proper?